src/mapgen_functions.cpp

```diff
@@ -977,382 +980,319 @@ void mapgen_fungal_flowers(map *m, oter_id, mapgendata dat, int, float)
     m->add_spawn(mon_fungaloid_seeder, 1, 12, 12);
 }
 
-void mapgen_road_straight(map *m, oter_id terrain_type, mapgendata dat, int turn, float density)
-{
-    bool sidewalks = false;
-    for (int i = 0; i < 8; i++) {
-        if (otermap[dat.t_nesw[i]].has_flag(has_sidewalk)) {
-            sidewalks = true;
+int terrain_type_with_suffix_to_nesw_array(oter_id terrain_type, bool array[4]) {
+    // extract the suffix from the terrain type name
+    std::string suffix = std::string(terrain_type).substr( std::string(terrain_type).find_last_of("_") + 1 );
+    // non-"_end" end tiles have _north _east _south _west, all of which contain "t"
+    if (suffix.find("t") != std::string::npos) {
+        suffix = suffix.substr(0,1);
+        if      (suffix=="n") suffix = "s"; // and they are all backwards :(
+        else if (suffix=="e") suffix = "w";
+        else if (suffix=="s") suffix = "n";
+        else if (suffix=="w") suffix = "e";
+    }
+    // manhole exception
+    if (suffix=="manhole") suffix="nesw";
+    // count and mark which directions the road goes
+    int num_dirs = 0;
+    if (suffix.find("n") != std::string::npos) { array[0] = true; num_dirs++; } else { array[0] = false; }
+    if (suffix.find("e") != std::string::npos) { array[1] = true; num_dirs++; } else { array[1] = false; }
+    if (suffix.find("s") != std::string::npos) { array[2] = true; num_dirs++; } else { array[2] = false; }
+    if (suffix.find("w") != std::string::npos) { array[3] = true; num_dirs++; } else { array[3] = false; }
```

Comment:
This code style is totally unlike most of the project. If anyone ever runs astyle over mapgen.cpp, it will get very long.
You could add a helper for it instead.